theano/gof/fg.py

```diff
@@ -478,8 +481,14 @@ def find_path_to(output_var, input_var):
 
         for node in new_nodes:
             assert node not in self.apply_nodes
-            self.__setup_node__(node)
-            self.apply_nodes.add(node)
+            prevent_addition = False
+            for n in self._removed_nodes :
+                if node is n :
+                    prevent_addition = True
+                    
+            if not prevent_addition : 
```

Comment:
Here I would keep the old code and if prevent_addiontion is true, I would raise an error of type InconsistencyError() with a messge like "Tring to reintroduce an old nodes in the graph. This should not happen"

I'm not sure what you are tring to do when we introduce an old node, but I think we should make the optimization that try that fail.